File: geoTweetsCrawler.rb
require 'tweetstream'
require 'json'
require 'mongo'
require 'time'

include Mongo

TweetStream.configure do |config|
  config.consumer_key       = ''
  config.consumer_secret    = ''
  config.oauth_token        = ''
  config.oauth_token_secret = ''
  config.auth_method        = :oauth
end


keyFilter = %w(id text lang media created_at user coordinates place entities)
userFilter = %w(id name screen_name lang location description statuses_count)
#puts keyFilter

Mongo::Logger.logger.level = Logger::WARN

# mongo_client = MongoClient.new("localhost", 27017)
  mongo_client = Mongo::Client.new([ '127.0.0.1:27017' ], :database => 'idea')
  coll = mongo_client[:geo_tweets_2]
  # coll = mongo_client['geo_tweets_2']

while true do
# begin

  dataList = []
  TweetStream::Client.new.locations(-180,-90,180,90) do |status|

  # The status object is a special Hash with
  # method access to its keys.
  #da(status.to_h)
    data = status.to_h
    data.select!{|key,value| keyFilter.include? key.to_s}
    data[:user].select!{|key,value| userFilter.include? key.to_s}
    data[:created_at] = Time.parse data[:created_at]

  #puts JSON.pretty_generate(data)
    dataList.push(data)
    #puts data
    # puts dataList.length
    if dataList.length > 100#10000
  	 # bulk = coll.initialize_ordered_bulk_op
     # bulk = coll('geo_tweets_2').initializeOrderedBulkOp()
     puts "超過100條記錄啦！"
     # puts bulk

  	 dataList.each do |data|
        coll.bulk_write([{ :insert_one => data}], :ordered => true)


  		  # bulk.insert(data)
  	 end
  	 # p bulk.execute
     puts "Insert"
     # count = 0
     # view = coll.find
     # view.each do |document|
     #      count += 1
     #      # p document
     #    end
     # p count
  	 dataList = []
    end

  
  end
rescue
  puts "Hey"
  sleep 10
end
end
#puts JSON.pretty_generate(x)
#rtc = retweet count id, text, lang, hashtags(emtity), media, created, screenname STC = "statuses_count,

File: geoTweetsCrawler.py
import sys
sys.path.append("/home/pan/Idealab/codes/twitterKeys")
import keys
from twython import TwythonStreamer
from pymongo import MongoClient
from datetime import datetime

class TweetStreammer(TwythonStreamer):

	def tweet_filter(self,data):

		keys = "id text lang media user coordinates place entities".split()
		user_keys = "id name screen_name lang location description statuses_count".split()
		tweet = {key: value for key, value in data.items() if key in keys }
		tweet["user"] = {key: value for key, value in data["user"].items() if key in user_keys }
		tweet["created_at"] = datetime.strptime(data["created_at"], '%a %b %d %H:%M:%S +0000 %Y')
		return tweet


	def __init__(self, consumer_key, consumer_secret, oauth_token, oauth_token_secret):
		super(TweetStreammer, self).__init__(consumer_key, consumer_secret, oauth_token, oauth_token_secret)
		self.tweets_buffer = []
		self.bulk_size = 1000
		self.collection = MongoClient("localhost", 27017)["twitter"]["geo_tweets"]

	def on_success(self, data):

		if "user" not in data:  #Skip the non-tweet request result
			return
		#print(data)
		tweet = self.tweet_filter(data)

		self.tweets_buffer.append(tweet)

		#print(tweet["coordinates"])
	
		if len(self.tweets_buffer) >= self.bulk_size:
			self.collection.insert(self.tweets_buffer)
			print("{} tweets have been inserted".format(len(self.tweets_buffer)))
			self.tweets_buffer = []


	def on_error(self, status_code, data):
		print(status_code)



consumer_key       = keys.CONSUMERKEY
consumer_secret    = keys.CONSUMERSECRET
oauth_token        = keys.OAUTHTOKEN
oauth_token_secret = keys.OAUTHTOKENSECRET




geo_stream = TweetStreammer(consumer_key, consumer_secret, oauth_token, oauth_token_secret)
while True:
	try:
		geo_stream.statuses.filter(locations = [-180,-90,180,90])  # Here indicate location of the tweet. It is also can be indicated null
	except:
		pass
